src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,282 @@
+                    AnalyzeSignatureType(paramLocation, param.Type);
+                }
+
+                void AnalyzeSignatureType(ImmutableArray<Location> locations, ITypeSymbol type)
+                {
+                    if (!type.IsUnmanagedType)
+                    {
+                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ManagedParameterOrReturnTypes));
+                    }
+
+                    if (type.IsValueType && TypeIsAutoLayoutOrContainsAutoLayout(type))
+                    {
+                        reportDiagnostic(locations.CreateDiagnostic(FeatureUnsupportedWhenRuntimeMarshallingDisabled, ByRefParameters));
+                    }
+                }
+            }
+
+            private bool TypeIsAutoLayoutOrContainsAutoLayout(ITypeSymbol type)
+            {
+                Debug.Assert(type.IsValueType);
+
+                try
+                {
+                    _layoutCacheLock.EnterReadLock();
+                    if (_isAutoLayoutOrContainsAutoLayoutCache.TryGetValue(type, out bool isAutoLayoutOrContainsAutoLayout))
+                    {
+                        return isAutoLayoutOrContainsAutoLayout;
+                    }
+                }
+                finally
+                {
+                    _layoutCacheLock.ExitReadLock();
+                }
+
+                if (_structLayoutAttribute is not null)
+                {
+                    foreach (var attr in type.GetAttributes(_structLayoutAttribute))
+                    {
+                        if (attr.ConstructorArguments.Length > 0
+                            && attr.ConstructorArguments[0] is TypedConstant argument
+                            && argument.Type is not null)
+                        {
+                            SpecialType specialType = argument.Type.TypeKind == TypeKind.Enum ?
+                                ((INamedTypeSymbol)argument.Type).EnumUnderlyingType.SpecialType :
+                                argument.Type.SpecialType;
+
+                            if (DiagnosticHelpers.TryConvertToUInt64(argument.Value, specialType, out ulong convertedLayoutKindValue) &&
+                                convertedLayoutKindValue == (ulong)LayoutKind.Auto)
+                            {
+                                try
+                                {
+                                    _layoutCacheLock.EnterWriteLock();
+                                    _isAutoLayoutOrContainsAutoLayoutCache.Add(type, true);
+                                    return true;
+                                }
+                                finally
+                                {
+                                    _layoutCacheLock.ExitWriteLock();
+                                }
+                            }
+                        }
+                    }
+                }
+
+                foreach (var member in type.GetMembers())
```

Comment:
Should we also be writing to the cache for true/false return values from here on in this method? It seems the cache is only ever populated with true values for structs with StructLayoutAttribute with auto layout kind, and we would end up repeatedly executing the code from here on in this method for structs without auto layout StructLayoutAttribute.